Libraries/MiscDrivers/SRAM/fastspi.c

```diff
@@ -0,0 +1,333 @@
+        *g_rx_buffer++ = SPI->fifo8[0];
+        g_rx_len--;
+    }
+
+    if (g_rx_len <= 0) {
+        g_rx_done = 1;
+    }
+
+    // Write any pending bytes out.
+    while (g_tx_buffer &&
+           (((SPI->dma & MXC_F_SPI_DMA_TX_LVL) >> MXC_F_SPI_DMA_TX_LVL_POS) < MXC_SPI_FIFO_DEPTH) &&
+           g_tx_len > 0) {
+        SPI->fifo8[0] = *g_tx_buffer++;
+        g_tx_len--;
+    }
+
+    if (g_tx_len <= 0) {
+        g_tx_done = 1;
+    }
+}
+
+void FastSPI_IRQHandler()
+{
+    uint32_t status = SPI->intfl;
+
+    if (status & MXC_F_SPI_INTFL_MST_DONE) { // Master done (TX complete)
+        g_master_done = 1;
+        SPI->intfl |= MXC_F_SPI_INTFL_MST_DONE; // Clear flag
+    }
+
+    if (status & MXC_F_SPI_INTFL_RX_THD) {
+        SPI->intfl |= MXC_F_SPI_INTFL_RX_THD;
+    }
+
+    if (status & MXC_F_SPI_INTFL_TX_THD) {
+        SPI->intfl |= MXC_F_SPI_INTFL_TX_THD;
+    }
+}
+
+int _dma_init()
+{
+    if (g_dma_initialized)
+        return E_NO_ERROR;
+
+    int err = MXC_DMA_Init();
+    // If we get a bad state error here it means DMA has already been
+    // initialized.  The drivers do a good job of handling this case,
+    // so it's safe to continue
+    if (err && (err != E_BAD_STATE))
+        return err;
+
+    g_tx_channel = MXC_DMA_AcquireChannel();
+    g_rx_channel = MXC_DMA_AcquireChannel();
+    if (g_tx_channel < 0 || g_rx_channel < 0) {
+        return E_NONE_AVAIL; // Failed to acquire DMA channels
+    }
+
+    // TX Channel
+    MXC_DMA->ch[g_tx_channel].ctrl &= ~(MXC_F_DMA_CTRL_EN);
+    MXC_DMA->ch[g_tx_channel].ctrl =
+        MXC_F_DMA_CTRL_SRCINC |
+        (0x2F
+         << MXC_F_DMA_CTRL_REQUEST_POS); // Enable incrementing the src address pointer, set destination to SPI0 TX FIFO (REQSEL = 0x2F)
+    MXC_DMA->ch[g_tx_channel].ctrl |=
+        (MXC_F_DMA_CTRL_CTZ_IE | MXC_F_DMA_CTRL_DIS_IE); // Enable CTZ and DIS interrupts
+    MXC_DMA->inten |= (1 << g_tx_channel); // Enable DMA interrupts
+
+    // RX Channel
+    MXC_DMA->ch[g_rx_channel].ctrl &= ~(MXC_F_DMA_CTRL_EN);
+    MXC_DMA->ch[g_rx_channel].ctrl =
+        MXC_F_DMA_CTRL_DSTINC |
+        (0x0F
+         << MXC_F_DMA_CTRL_REQUEST_POS); // Enable incrementing the dest address pointer, set to source to SPI0 RX FIFO (REQSEL = 0x0F)
+    MXC_DMA->ch[g_rx_channel].ctrl |=
+        (MXC_F_DMA_CTRL_CTZ_IE | MXC_F_DMA_CTRL_DIS_IE); // Enable CTZ and DIS interrupts
+    MXC_DMA->inten |= (1 << g_rx_channel); // Enable DMA interrupts
+
+    MXC_NVIC_SetVector(MXC_DMA_CH_GET_IRQ(g_tx_channel), DMA_TX_IRQHandler);
+    NVIC_EnableIRQ(MXC_DMA_CH_GET_IRQ(g_tx_channel));
+    NVIC_SetPriority(MXC_DMA_CH_GET_IRQ(g_tx_channel), 0);
+
+    MXC_NVIC_SetVector(MXC_DMA_CH_GET_IRQ(g_rx_channel), DMA_RX_IRQHandler);
+    NVIC_EnableIRQ(MXC_DMA_CH_GET_IRQ(g_rx_channel));
+    NVIC_SetPriority(MXC_DMA_CH_GET_IRQ(g_tx_channel), 0);
+
+    g_dma_initialized = true;
+
+    return E_NO_ERROR;
+}
+
+int spi_init()
+{
+    MXC_SYS_ClockEnable(MXC_SYS_PERIPH_CLOCK_SPI0);
```

Comment:
We should figure out a scheme to avoid hardcoding these MXC_SYS defines.

Reply:
There are other hard-coded settings in these drivers specific to the FTHR and the N01S.  I put some todos to tackle this when we add support for other srams